src/defineFunction.js

```diff
@@ -1,94 +1,127 @@
-        // The default value is `1`
-        greediness?: number,
-
-        // Whether or not the function is allowed inside text mode
-        // (default false)
-        allowedInText?: boolean,
-
-        // Whether or not the function is allowed inside text mode
-        // (default true)
-        allowedInMath?: boolean,
-
-        // (optional) The number of optional arguments the function
-        // should parse. If the optional arguments aren't found,
-        // `null` will be passed to the handler in their place.
-        // (default 0)
-        numOptionalArgs?: number,
-
-        // Must be true if the function is an infix operator.
-        infix?: boolean,
-    },
+    props: FunctionPropSpec,
 
     // The handler is called to handle these functions and their arguments.
-    // It receives two arguments:
-    //   - context contains information and references provided by the parser
-    //   - args is an array of arguments obtained from TeX input
-    // The context contains the following properties:
-    //   - funcName: the text (i.e. name) of the function, including \
-    //   - parser: the parser object
-    //   - lexer: the lexer object
-    // The latter three should only be used to produce error messages.
     //
     // The function should return an object with the following keys:
     //   - type: The type of element that this is. This is then used in
     //          buildHTML/buildMathML to determine which function
     //          should be called to build this node into a DOM node
     // Any other data can be added to the object, which will be passed
     // in to the function in buildHTML/buildMathML as `group.value`.
-    handler: (context: any, args: any) => T,
+    handler: ?FunctionHandler,
 
     // This function returns an object representing the DOM structure to be
     // created when rendering the defined LaTeX function.
-    htmlBuilder: (group: T, options: Options) => any,
+    // TODO: Port buildHTML to flow and make the group and return types explicit.
+    htmlBuilder?: (group: *, options: Options) => *,
 
     // This function returns an object representing the MathML structure to be
     // created when rendering the defined LaTeX function.
-    mathmlBuilder: (group: T, options: Options) => any,
-}
+    // TODO: Port buildMathML to flow and make the group and return types explicit.
+    mathmlBuilder?: (group: *, options: Options) => *,
+|};
+
+/**
+ * Final function spec for use at parse time.
+ * This is almost identical to `FunctionPropSpec`, except it
+ * 1. includes the function handler, and
+ * 2. requires all arguments except argTypes.
+ * It is generated by `_defineFunction`.
+ */
+type FunctionSpec = {|
+    numArgs: number,
+    argTypes?: ArgType[],
+    greediness: number,
+    allowedInText: boolean,
+    allowedInMath: boolean,
+    numOptionalArgs: number,
+    infix: boolean,
+    // Must be specified unless it's handled directly in the parser.
+    handler: ?FunctionHandler,
```

Comment:
Thanks for calling this out here.  I think we should try to move the handlers for implicit group functions into the function definition.  I opened #868 for this purpose.

Reply:
SGTM! That would be much more satisfying.
That being said, spreading out parsing responsibilities a lot can get confusing, so as things become clearer.